chronos/asyncfutures2.nim

```diff
@@ -60,6 +60,10 @@ type
       closure*: iterator(f: Future[T]): FutureBase {.raises: [Defect, CatchableError, Exception], gcsafe.}
     value: T ## Stored value
 
+  # Future with a tuple of possible exception types
+  # eg FuturEx[void, (ValueError, OSError)]
+  FuturEx*[T, E] = Future[T]
```

Comment:
Does this essentially creates an alias instead of a new type? It is a bit of a puzzling construct.

Reply:
This construct cannot produces a reliable solution indeed. You cannot count that the compiler won't eliminate the information about `E` during the generic instantiation process. Nim tries very hard to normalize generic alias types to their ultimate destination type in `semtypinst` in the compiler.

Reply:
To clarify further, defining this as a distinct type would work, but what would be the implications of this?

Reply:
As long as this snippet works, FuturEx will work: https://play.nim-lang.org/#ix=3LXl
And it would seem weird to me if it didn't work

Basically, FuturEx sole purpose in life is to give CT information to await that this future can raise X exceptions

Having a distinct type would probably create havoc in some way, require converter or borrows which are not great

Reply:
I'm explaining that the compiler will happily erase that information behind your back. You really cannot count on it unless you make your `Gen[T]` type a `distinct int`. You can trust me because this is the area of the compiler that I have most heavily contributed to, but if you insist I can construct some more complex scenarios that will break your example.

Reply:
There you go:

```nim
type Gen[G] = object
type Gen2[G, E] = Gen[G]

proc test: Gen2[bool, int] = discard

proc checkTestReturnType(T: type, checkMsg: string) =
  let x = test()
  assert type(x) is T, checkMsg
  
checkTestReturnType Gen2[bool, float], "this should be true"
checkTestReturnType Gen2[bool, int], "this should be false"
checkTestReturnType Gen2[string, int], "this shouldn't be reached"
```

Fails with "this shouldn't be reached"

Reply:
I trust you, but need to understand ;)

I do expect him to erase information when converting a `FuturEx` to a `Future`, since we need implicit conversion in that case, but indeed, he seems to erase in more cases than I would like:
```nim
let a: FuturEx[void, (CatchableError,)] = FuturEx[void, (Defect,)].new # compiles, which is not good
# not consistent with
type Gen[T] = int
# let a: Gen[float] = Gen[bool](5) # doesn't compile
type Gen2[T, E] = Gen[T]
let b: Gen2[bool, float] = Gen2[bool, bool](5) # compiles, which is not good
```
So, a "useless" generic parameter pointing to a concrete type will be checked, but not a "useless" generic parameter pointing to a generic type, afaict?

This will cause issue if you do
```nim
proc test1 {.async, raises: [].} = discard
proc test2 {.async, raises: [ValueError].} = raise newException(ValueError, "surprise!")
var fut = test1()
await fut
fut = test2() # should be illegal cause fut kept his type
await fut # will throw defect
```

So that's pretty bad. Now, as I said, I don't see a way to make it work with distinct or others. To keep backward compatibility, we need an implicit FuturEx -> Future conversion, but a locked Future -> FuturEx conversion.

So, we have a few options:
- We find a way to improve that, which would be ideal. The inconsistency makes me thing we could say it's a nim issue? Or if we find a solution to make it distinct or other
- We keep it as-is, and warn about misusages. I think it's reasonable because it fails on "advanced" use cases, unless we found other ways it fail
- We say goodbye to backward compatibility
- We say goodbye to checked exceptions

Personally, now that I've tested a bit of checked exceptions in libp2p with this branch, I can't think going without them.

Reply:
Ok github doesn't seem to auto-update the comments anymore, nice.. So I guess we've hit the same issue independently :)

Reply:
That's not a Nim issue. A type alias is a statement that two types are equivalent. The code below is correct, because:

```nim
let a: FuturEx[void, (CatchableError,)] = FuturEx[void, (Defect,)].new # compiles, which is not good
# is the same as
let a: Future[void] = Future[void].new
```

You can say goodbye to backwards compatibility or you can go with my suggested approach of encoding the information with an additional type:

```
proc someAsync(): Future[Raising[T, (ValueError, KeyError)]
```

`await` can be specilized to skip over the `raising` type if desired.

Reply:
Interesting suggestion, `Raising` indeed seems like a good meddle ground. I wonder why `Future[T, E = (CatchableError)]` wouldn't work tho?

Reply:
@zah would
```nim
type
  FuturEx[T, E] = ref object of FutureBase
     [...]
  Future[T] = FuturEx[T, (CatchableError,)]
```
Be OK? I've surprisingly managed to make it work with few adjustments, which indicates me that it probably also causes issue. But intuitively, I don't see how the nim compiler could throw away my types if it's part of the "source" of the alias.
At first glance, it seems pretty ok (the example I gave earlier no longer compiles).

Otherwise, need to think about the Raising type

Reply:
Ah, thats even better - hopefully it works. Worst case, `Raising` would be viable path forward IMO.

Reply:
> which indicates me that it probably also causes issue

I'm speculating, but it looks like a more backwards compatible approach - `Future[T] = FuturEx[T, (CatchableError,)]` is essentially exactly the same as `Future[T] = ref object of FutureBase`, because `E` gets erased in most cases? 

If that is indeed the case, I think it's a safer overall approach and we can upgrade the codebase piecemeal without introducing unexpected breakage. One issue I think the current approach has is that it essentially changes the shape of the type underneath and could potentially introduce runtime issues?

Reply:
Current as in the `Future[T, E] = Future[T]`

Reply:
nvm, this doesn't have any runtime effect whatsoever.

Reply:
The proposed `Future[T] = FuturEx[T, (CatchableError,)]` definition wouldn't suffer from the problems I described indeed. If you managed to make it work, that's great.